server/src-lib/Hasura/GraphQL/Execute.hs

```diff
@@ -25,13 +36,68 @@ import           Hasura.HTTP
 import           Hasura.Prelude
 import           Hasura.RQL.DDL.Headers
 import           Hasura.RQL.Types
+import           Hasura.SQL.Types
 
+import qualified Hasura.GraphQL.Resolve.ContextTypes    as CT
 import qualified Hasura.GraphQL.Validate                as VQ
+import qualified Hasura.GraphQL.Validate.Field          as VF
 import qualified Hasura.GraphQL.Validate.Types          as VT
 
+data JoinInfo
+  = JoinInfo
+  { jiParentAlias   :: !G.Alias
+  , jiParentJoinKey :: !G.Name -- change to alias
+  , jiChildJoinKey  :: !G.Name
+  , jiChildAlias    :: !G.Alias
+  , jiNamespace     :: !(Maybe G.Name)
+  } deriving (Show, Eq)
+
+newtype GraphQLRequestPromise
+  = GraphQLRequestPromise {
+      grpResolve  :: forall m. (MonadError QErr m) => [J.Value] -> m GraphQLRequest
+    }
+
+instance Show GraphQLRequestPromise where
+  show _ = "GraphQLRequestPromise Func"
+
+data GExPDepRemote = GExPDepRemote !RemoteSchemaInfo !GraphQLRequestPromise JoinInfo IsAsync
+  deriving Show
+
+data GExPHasura = GExPHasura !GCtx !VQ.RootSelSet [GExPDepRemote]
+  deriving Show
+
+data GExPRemote
+  = GExPRemote
+  { _geprRemoteInfo :: !RemoteSchemaInfo
+  , _geprRequest    :: !GraphQLRequest
+  , _geprIsAsync    :: !IsAsync
+  } deriving (Show, Eq)
+
 data GQExecPlan
-  = GExPHasura !GCtx !VQ.RootSelSet
-  | GExPRemote !RemoteSchemaInfo !G.TypedOperationDefinition
+  = GQExecPlan
+  { _gqepHasura :: !(Maybe GExPHasura)
+  , _gqepRemote :: ![GExPRemote]
+  , _gqepOpType :: !G.OperationType
+  } deriving Show
```

Comment:
This execution plan has the set of remotes, and maybe a Hasura database -- but if there is no mention of an entity from the Hasura PG schema, it will be `Nothing`, and the remotes didn't depend on any JSON values from it. In any case, it's ready to run.